### PR TITLE
Add slice support for *_argnum parameters

### PR DESCRIPTION
**Pull request by dipampaul17**
# Implement Slice Support for `*_argnum` Parameters

This PR implements support for slice objects in JAX functions that accept `*_argnum` parameters, addressing issue #28667.

## What's Changed

1. Modified the core utility functions in `jax/_src/api_util.py` that handle argnum parameters:
   - Updated `_ensure_index` and `_ensure_index_tuple` to handle slice objects
   - Added comprehensive validation and error handling for slice parameters
   - Maintained backward compatibility with existing code

2. Added comprehensive tests in `jax/_src/test_argnum_slice.py` to verify:
   - Core utility functions with various slice patterns
   - Integration with JAX transformations (grad, value_and_grad, jacfwd, jit)
   - Edge cases (negative indices, None values, etc.)
   - Proper error handling for invalid inputs

## Working Transformations

We've verified that slice support works with the following JAX transformations:

1. ✅ `jax.grad`: Takes derivatives with respect to a range of arguments
2. ✅ `jax.value_and_grad`: Returns both value and gradients for a range of arguments
3. ✅ `jax.jacfwd`: Computes Jacobians for multiple arguments at once
4. ✅ `jax.jit`: Specifies a range of static arguments

## Documentation Updates

The following docstring updates should be made to document the new slice support:

1. In `jax/_src/api.py` - `grad` function:
```python
argnums: Optional, integer, sequence of integers, or slice object. Specifies which
  positional argument(s) to differentiate with respect to (default 0).
  If a slice is provided (e.g., ``slice(0, 2)``), it will be converted to a tuple
  of indices using Python's range function. Slice stop values must be specified.
```

2. Similar updates for `value_and_grad`, `jacfwd`, `jacrev`, `jit`, etc.

### Basic Usage

Before:
```python
# To differentiate with respect to the first two arguments
grad_f = jax.grad(f, argnums=(0, 1))
```

After:
```python
# To differentiate with respect to the first two arguments
grad_f = jax.grad(f, argnums=slice(0, 2))
```

### Advanced Usage

```python
# Differentiating with respect to arguments 0 and 2 (skipping 1)
def f(x, y, z):
    return x * y * z
grad_f = jax.grad(f, argnums=slice(0, 3, 2))
grad_x, grad_z = grad_f(2.0, 3.0, 4.0)
# grad_x = 12.0 (= y*z), grad_z = 6.0 (= x*y)

# Using negative indices
def g(w, x, y, z):
    return w * x * y * z
grad_g = jax.grad(g, argnums=slice(-4, -2))
grad_w, grad_x = grad_g(2.0, 3.0, 4.0, 5.0)
# Equivalent to slice(0, 2)

# Using None as start (defaults to 0)
grad_h = jax.grad(g, argnums=slice(None, 2))
# Equivalent to slice(0, 2)

# Multiple transformations with slice syntax
def f(x, y, z):
    return x * y * z

# Get both value and gradients
val_grad_f = jax.value_and_grad(f, argnums=slice(0, 2))
value, (grad_x, grad_y) = val_grad_f(2.0, 3.0, 4.0)

# Jacobian computation for multiple arguments
jac_f = jax.jacfwd(f, argnums=slice(0, 2))
jac_x, jac_y = jac_f(2.0, 3.0, 4.0)

# Static arguments for jit
@jax.jit(static_argnums=slice(0, 2))
def g(a, b, c, d):
    # a and b will be treated as static
    return a * b * c * d
```

## Design Decisions

1. **Requiring explicit `stop` value**: To prevent unbounded slices that could lead to unexpected behavior, the implementation requires that `stop` is explicitly specified (i.e., not `None`).

2. **Default handling**: 
   - When `start` is `None`, it defaults to 0 (following Python slice convention)
   - When `step` is `None`, it defaults to 1 (following Python slice convention)

3. **Validation of slice outputs**: The implementation validates that the slice produces at least one index, raising an informative error if an empty sequence would be produced (e.g., `slice(2, 1)`).

4. **Comprehensive error messages**: Clear and specific error messages are provided for various invalid slice specifications to aid debugging.

## Testing

Tests in `jax/_src/test_argnum_slice.py` verify:
1. Basic functionality with different slice patterns
2. Edge cases including negative indices, None values, and full-range slices
3. Error handling for invalid inputs
4. Integration with multiple JAX transformations

## Performance Considerations

The implementation has negligible performance impact since it only affects the processing of function arguments before tracing/compilation, not the actual computation.

## Related Issues

Resolves #28667

### CODE REVIEW

**Comment by jakevdp** [comment]
Closing due to lack of CLA, and also because this looks LLM-generated.
